src/libaktualizr/bootloader/bootloader.cc

```diff
@@ -169,13 +174,26 @@ void Bootloader::rebootFlagClear() {
 }
 
 void Bootloader::reboot(bool fake_reboot) {
+  size_t len = reboot_command_.length();
+  struct stat st;
+  char *cmd = NULL, full_cmd[256];
+
   if (fake_reboot) {
     boost::filesystem::remove(reboot_sentinel_);
   } else {
-    sync();
     if (setuid(0) != 0) {
       LOG_ERROR << "Failed to set/verify a root user so cannot reboot system programmatically";
+      return;
+    }
+
+    reboot_command_.copy(full_cmd, len);
+    full_cmd[len] = '\0';
+    cmd = strtok(full_cmd, "\t ");
+    if ((stat(cmd, &st) >= 0) && (st.st_mode & (S_IXUSR | S_IXGRP | S_IXOTH))) {
+      system(reboot_command_.c_str());
     } else {
+      LOG_WARNING << "The preferred reboot command: '" << reboot_command_ << "' does not exist or is not executable";
+      sync();
       ::reboot(RB_AUTOBOOT);
```

Comment:
IMHO, system("reboot") (it will call 'systemd reboot' in our default case) would work in most of cases so I would use it as a default value in case if 'reboot_command' is not specified in the config. Not sure if it makes to fallback to it if the 'reboot_command' fails - agree with lbonn here.
Also, it looks a bit insecure/dangerous to allow an user to execute whatever command they specify in the config, it could be 'rm -rf /', 'cat /etc/passwd', etc :)

Reply:
aktualizr runs as root and should only read configurations writable by root + it does dangerous operations by nature. Using a hard-coded list sounds like a band-aid that can be bypassed.

A possible security check would be that the executable indeed belongs to root and is not writeable by group and others.

Reply:
I don't think we should add another way for user to shoot themselves in a foot. How many ways to reboot a sytem do we really have to support? Also, those commands are very rarely change so we can simply use "reboot_command = systemv" and use /sbin/reboot explicitly in this case instead of default "systemctl reboot". If we really have to we can always expand this by adding "reboot_command = systemwhatever" which will map to another command to add support for some exotic system. Using arbitrary command is more flexible of course but I don't think it's worth it in this case.

Reply:
In the end it doesn't make any difference, if we make room for a calling an arbitrary script with a fixed name, we can as well allow an arbitrary command. It doesn't add nor subtract any foot-gun.

My wish was primarily to allow this use case: https://github.com/toradex/meta-toradex-torizon/blob/a8b50aa8b9f7f14eb60ebc8fc68919ba87413804/recipes-sota/ostree/files/ostree-pending-reboot.path. Touching a file as a trigger to a systemd service seems really useful as it allows a variety of use cases which need close interaction with the init system.
See here for background: https://github.com/advancedtelematic/meta-updater/pull/486.

In this case, the reboot command would just be `touch /run/aktualizr-session/trigger_reboot`

Reply:
That make sense, nice. I wonder if we could always use reboot via this external path trigger.